src/nunavut/lang/c/support/nunavut.h

```diff
@@ -0,0 +1,461 @@
+/*
+ *
+ * UAVCAN serialization support routines.                                                                     +-+ +-+
+ *                                                                                                            | | | |
+ *                                                                                                            \  -  /
+ *  This software is distributed under the terms of the MIT License.                                            ---
+ *                                                                                                               o
+ * +------------------------------------------------------------------------------------------------------------------+
+ */
+#ifndef NUNAVUT_SUPPORT_SERIALIZATION_H_INCLUDED
+#define NUNAVUT_SUPPORT_SERIALIZATION_H_INCLUDED
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <string.h>
+#include <float.h>
+#include <assert.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+/* TODO: issue #112, add serialization support for C types. */
+typedef uint32_t NunavutUnsignedBitLength;
+typedef int32_t NunavutSignedBitLength;
+typedef float NunavutFloat32;
+typedef double NunavutFloat64;
+
+// appropriate name?
+// do we typedef this, leave it anon, or enum NAME
+enum NUNAVUT_ERR
+{
+    NUNAVUT_ERR_INVALID_BUF         = -1,
+    NUNAVUT_ERR_INVALID_LEN         = -2,
+    NUNAVUT_ERR_INVALID_OFFSET      = -3,
+    NUNAVUT_ERR_INVALID_TAG         = -4,
+};
+
+/// There are two implementations of the primitive (de-)serialization algorithms: a generic one, which is invariant
+/// to the native byte order (and therefore compatible with any platform), and the optimized one which is compatible
+/// with little-endian platforms only. By default, the slow generic algorithm is used.
+/// If the target platform is little-endian, the user can enable this option to use the optimized algorithm.
+#ifndef NUNAVUT_CONFIG_LITTLE_ENDIAN
```

Comment:
macro use is undesirable for high reliability software. Can we, instead, have two headers; one for LE and another for BE that provide only the implementations where these diverge. The include used can then be generated from templates driven by nunavut configuration options.

Reply:
> one for LE and another for BE

Probably you mean one for LE and one for any byte order. The current implementation has a default implementation that suits all platform and an optimized option that is only suitable for LE.